Lombiq.HelpfulLibraries.OrchardCore/ResourceManagement/ResourceFilterMiddleware.cs

```diff
@@ -1,43 +1,63 @@
-using Microsoft.AspNetCore.Http;
-using Microsoft.Extensions.DependencyInjection;
-using OrchardCore.ResourceManagement;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading.Tasks;
-
-namespace Lombiq.HelpfulLibraries.OrchardCore.ResourceManagement;
-
-public class ResourceFilterMiddleware
-{
-    private readonly RequestDelegate _next;
-
-    public ResourceFilterMiddleware(RequestDelegate next) => _next = next;
-
-    public async Task InvokeAsync(HttpContext context)
-    {
-        var builder = new ResourceFilterBuilder();
-        var anyProviders = context
-            .RequestServices
-            .GetService<IEnumerable<IResourceFilterProvider>>()
-            .ForEach(provider => provider.AddResourceFilter(builder));
-
-        if (anyProviders)
-        {
-            IResourceManager resourceManager = null;
-
-            var activeFilters = await builder
-                .ResourceFilters
-                .Where(filter => filter.Filter != null || filter.FilterAsync != null)
-                .WhereAsync(
-                    filter => filter.Filter != null
-                        ? Task.FromResult(filter.Filter(context))
-                        : filter.FilterAsync(context));
-
-            activeFilters.ForEach(
-                filter => filter.Execution(resourceManager),
-                _ => resourceManager = context.RequestServices.GetService<IResourceManager>());
-        }
-
-        await _next(context);
-    }
-}
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Admin;
+using OrchardCore.ResourceManagement;
+using OrchardCore.Themes.Services;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Lombiq.HelpfulLibraries.OrchardCore.ResourceManagement;
+
+public class ResourceFilterMiddleware
+{
+    private readonly RequestDelegate _next;
+
+    public ResourceFilterMiddleware(RequestDelegate next) => _next = next;
+
+    public async Task InvokeAsync(HttpContext context)
+    {
+        var providers = context
+            .RequestServices
+            .GetRequiredService<IEnumerable<IResourceFilterProvider>>()
+            .Select(provider => new { Provider = provider, ThemeRequirements = provider.GetRequiredThemes().ToList() })
+            .ToList();
+
+        IList<string> themes =
+            providers.Any(providerInfo => providerInfo.ThemeRequirements.Any())
+                ? new[]
+                    {
+                        await context.RequestServices.GetRequiredService<ISiteThemeService>().GetSiteThemeAsync(),
+                        await context.RequestServices.GetRequiredService<IAdminThemeService>().GetAdminThemeAsync(),
```

Comment:
Is my understanding correct that these two invocations run in sequence? How would you run them in parallel?

Reply:
Why would you want that?

Reply:
To speed up execution. But your question already makes me think that that's not the reason for using `async/await` - it's rather not to congest the thread.

Reply:
Yep, and within the context of an ASP.NET Core HTTP request it rarely makes sense to parallelize, and if it does then you should take extra care. Here, neither call is guaranteed to be thread-safe, but actually, both will almost always come from a cache and be trivial.